### PR TITLE
Fix setState to work in callback form and to allow partial intellisense

### PR DESCRIPTION
**Pull request by ericanderson**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

See: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18365#issuecomment-351013257

### CODE REVIEW

**Review by vsiao — CHANGES_REQUESTED** [review]
Mind adding some `// $ExpectError` tests for this? Looks good otherwise.

**Comment by ericanderson** [comment]
@vsiao GTG

File: types/react/test/tsx.tsx [hunk 1]
```diff
@@ -77,3 +77,20 @@ const StatelessComponentWithoutProps: React.SFC = (props) => {
         </React.Fragment>
     </React.Fragment>
 </div>;
+
+class Comp extends React.Component<{}, { foo: boolean, bar: boolean }> {
```
Comment (vsiao) [inline types/react/test/tsx.tsx]: Maybe could name this component `SetStateTest`? :)

**Review by vsiao — APPROVED** [review]
Thanks for the thorough suite of tests!

**Comment by johnnyreilly** [comment]
Thanks!

**Pull request merged**